feat: new blogpost (#314)
satellytes/satellytes.com: blog-posts/scoped-registry.md

---
path: "/blog/scoped-registry/"
date: "2021-12-30"

> title: "NPM scoped registry and authentication"

Reviewer: i think this title is okay but could be catchier. Did you have any alternatives in mind? I'm thinking of more telling titles like "The basic of different npm registries?" or "Why you should know npm scopes". You current title is very non telling, could be anything about npm registries

Author: You are right. I tried to make it more descriptive

featuredImage: images/npm-scoped-registry.jpg
attribution:
    creator: Paul Esch-Laurent

>     source: https://unsplash.com/photos/oZMUrWFHOB4

Reviewer: Thanks for a beautiful image and respecting the attribution 👍

Author: :+1:

seoMetaText: create scoped packages and scoped registry and configure authentication via npmrc

> teaserText: What are scoped packages, scoped registry and configure authentication via npmrc

Reviewer: Did you know where the text end up? Seo is for search engines. The teaser text is for the blog overview. I wonder if this is understandable or what you thought about that?

Author: I improved the texts. Please have another look. :pray:

author: Mark Altmann

> authorSummary: dev@satellytes, ❤️ for Wombats

Reviewer: I like this and this is what I want to see, unique and personal touches in a blog post 💛

---

> ## TL;DR

Reviewer: I really don't like TL;DR. You invest so much time into a blog post and then you want to let the user read a summary before going into the topic? I think that's the worst story telling approach. You can put that part at the end with the same headline. People who are interested in those summaries will find it. Actually I would prefer a "Summary". `TL;DR` being "Too long, didn't read"  is a really disrespectful phrase for me showing more disinterest then interest in a text.

Author: I removed the TL;DR; and put it into a conclusion and connected it with the story which makes the story smoother. 
But in general I think it depends on what you focus and what the people want to see. Personally when I am on my way to find a solution for a problem I am very grateful for TL;DR; because I want a solution or some piece of code and not pieces scattered over a story I have to read. On the other side when I like to read the story I could go on with the story to read and if I learn something in the end even better. But this is really a personal opinion I think. :+1:

Reviewer: I understand and thanks for the changes 🙌

Packages in npm can be grouped together with scopes. Scopes are preceded by a `@` followed by the scope name and the package name separated by `/` e.g. `@scope/package`. To set a specific registry for a scope you can use scoped registries. To set a scoped registry open up the .npmrc file and enter the following:

> `@scope:registry=https://<your registry>` where `@scope` is your scope name and `<your registry>` is the registry where your scoped packages are hosted. 

Reviewer 2: To make it more readable I would put this into a full-width code block (three ` ticks). This way you can see on first glance what the file should look like:

```
@scope:registry=https://<your registry>
```

Author: good idea. Thanks :+1:

If you need authentication you can add an additional line below the scoped registry with `_auth` and the basic authentication string for username and password (`username:password` ⇒ base64 encoded) or `_authToken` for an access token:

> `//<your registry>:_auth=<base64-string>`

Reviewer: Is this was you want to transport with this blog post. Kind of a "how to use a different npm registry" with the added insight why scopes are so useful within that context?

Author: What do you mean exactly? I want to show how to use more than one registries (with the use of scoped registries) and in the end how to access them with credentials.

Reviewer: I think it's clear now with the added text that's why my question looks weird ( I don't understand it either anymore)

> `//<your registry>:_auth=<base64-string>`

Reviewer 2: Also here it would be simpler to have it full width and with the first line.

```
@scope:registry=https://<your registry>
//<your registry>:_auth=<base64-string>
```

To be consistent with the example scope maybe it's worth also adding an example without placeholders:

```
@scope:registry=https://registry.wombatcorp.org
//registry.wombatcorp.org/:_auth=c2lyd29tYmF0OnRoZWN1ZGQxZXIh
```

Author: :+1: :pray:

## What is this all about?

> Nowadays almost every programming language comes with its own package management solution, so developers can easily share packages with other developers around the world. For Python it is pip, for .NET it`s nuget. npm is the package manager solution for Javascript based languages. It is used to create and use node packaged modules and is built into the Javascript platform [Node.js](http://www.nodejs.org/). The central component behind these package managers is a registry. A registry is a database of packages, each comprised of software and metadata. For example the public registry for npm is [registry.npmjs.org](http://registry.npmjs.org).

Reviewer 2: You are using the wrong apostrophe here:

it`s nuget -> it's nuget

Author: :+1:

## Scopes and scoped registries

If you are a javscripty/typescript developer in a company you certainly stumbled over the problem that you, on the one side, use official packages from the public npm registry but on the other side also need internal packages from the company internal registry. Unfortunately there cannot be two main registries in npm at the same time. But scoped packages come to the rescue.

> All npm packages have a name. Some package names also have a scope. A scope follows the usual rules for package names (URL-safe characters, no leading dots or underscores). When used in package names, scopes are preceded by an `@` symbol and followed by a slash, e.g. `@somescope/somepackagename` Scopes are a way of grouping related packages together, and also affect a few things about the way npm treats the package.
> -- <cite>[npm](https://docs.npmjs.com/cli/v8/using-npm/scope)</cite>

Packages can be grouped together with scopes and for scopes we can set alternative registries named `scoped registries` where npm will lookup the packages instead of the main registry.

> > 💡 The usage of scoped packages and scoped registries can also greatly diminish your attack surface for the dependency confusion attack [dependency confusion attack](https://snyk.io/blog/detect-prevent-dependency-confusion-attacks-npm-supply-chain-security/)

Reviewer: I like thos small bits of knowledge you placed 👍

Author: I thought it might be of interest with the recent topics and also another reason not to use a proxy repository which is possible in e.g. nexus :smile:

## .npmrc

npmrc is a configuration file configures how npm works in your environment. It can be used globally or on a user or project level. [Here](https://docs.npmjs.com/cli/v8/configuring-npm/npmrc) are the docs for npmrc.

> Lets suppose we have a company named *Wombat Corp.* and have published our packages under the scope `@wombatcorp`. Our registry can be found under the url `registry.wombatcorp.org`. We could add a scoped registry with the following line:

Reviewer: you use `wombatcorp.org` where I think `example.com` as this will never be owned by some individual. On the other hand I see how to want to tell a story around wombat here which is fine for me 🙌

Author: I changed it to Working Wombats. This is more generic and should be no problem I hope :smile:

`@scope:registry=https://<your registry>`

So to add our example registry we have to add the following line to npmrc:

`@wombatcorp:registry=https://registry.wombatcorp.org/`

> ## 401 Unauthorized... what 😱?

Reviewer 2: I like the emojis but would put them after the question mark as it could degrade accessibility otherwise.

```
## 401 Unauthorized... what? 😱
```

Author: also very good. thank you. :+1:

---
## What is this all about?

Nowadays almost every programming language comes with its package management solution, so developers can easily share packages with other developers around the world. For Python it is pip, for .NET it's NuGet. npm is the package manager solution for JavaScript-based languages. It is used to create and use node packaged modules and is built into the JavaScript platform [Node.js](http://www.nodejs.org/). The central component behind these package managers is a registry. A registry is a database of packages, each comprised of software and metadata. For example, the public registry for npm is [registry.npmjs.org](http://registry.npmjs.org).

Besides using public registries, companies can establish their private registries in their company network. [Nexus](https://www.sonatype.com/products/repository-pro), [Artifactory](https://jfrog.com/artifactory/) or [verdaccio](https://github.com/verdaccio/verdaccio) only to name some of the well-known. The advantage is that the published packages of that company never leave the company network. Another advantage is that you can set up authentication to additionally secure your internally published packages.

> ## Real-world challenges for a fictional enterprise

Reviewer: love it 🔥

Author: :100:

**Use environment variables**

we could export the authentication string as an environment variable:

e.g. for linux `$ export NPM_TOKEN=<base64-string>`

We can then add the env variable in our .npmrc file:

```
//registry.wombatcorp.org/:_auth=${NPM_TOKEN}
```

The npm CLI will replace this value with the contents of the NPM_TOKEN environment variable
    
**Add scoped registry to user-level instead of project-level**

There are different .npmrc files as explained here ([npmrc](https://docs.npmjs.com/cli/v8/configuring-npm/npmrc)). 
Each of these files is loaded, and config options are resolved in priority order. For example, a setting in the user config file would override the setting in the global config file. So we could put the scoped registry into the user config file instead of the project config file.

## Conclusion

Finally, Jeff could shut down his computer and leave work. Today he learned how to use scoped registries to access different scoped packages. He also learned how to use .npmrc for configuration, what Basic Authentication is and how to use it to access scoped registries.

"In the end...", thought Jeff, "...this wasn't too hard."

> As a backend developer myself peeking into the world of frontend, I was standing before the same problem as Jeff. I wish I had found this blog post by then... 😌

Reviewer: That is the perfect conclusion ending, I love it

Author: :fire:

featuredImage: images/npm-scoped-registry.jpg
attribution:
    creator: Paul Esch-Laurent
    source: https://unsplash.com/photos/oZMUrWFHOB4
seoMetaText: Access protected registries with npm scoped packages and scoped registries
teaserText: If you use multiple registries (public and private) you can use npm and scoped registries to configure access to multiple registries even if they are protected.
author: Mark Altmann
authorSummary: dev(ops)@satellytes, ❤️ for Wombats

---
## What is this all about?

Nowadays almost every programming language comes with its own package management solution, so developers can easily share packages with other developers around the world. For Python it is pip, for .NET it's nuget. npm is the package manager solution for Javascript based languages. It is used to create and use node packaged modules and is built into the Javascript platform [Node.js](http://www.nodejs.org/). The central component behind these package managers is a registry. A registry is a database of packages, each comprised of software and metadata. For example the public registry for npm is [registry.npmjs.org](http://registry.npmjs.org).

> Beside using public registries, companies can establish their own private registries in their company network. Nexus, Artifactory or verdaccio only to name some of the well-known. The advantage is that the published packages of that company never leave the company networks. Another advantage is that you can setup authentication to additionally secure your internally published packages.

Reviewer: As a reader, who doesn't know let's say nexus, wouldn't you love to click them and check them out? Link them all 🙌 For your convenience a suggestion to merge:

```suggestion
Beside using public registries, companies can establish their own private registries in their company network. [Nexus](https://www.sonatype.com/products/repository-pro), [Artifactory](https://jfrog.com/artifactory/) or [verdaccio](https://github.com/verdaccio/verdaccio) only to name some of the well-known. The advantage is that the published packages of that company never leave the company networks. Another advantage is that you can setup authentication to additionally secure your internally published packages.
```

Author: Thanks for that. Thats a really good idea with the links.

> Jeff Wombat Jr.[^1] is a javscripty/typescript developer at **"Working Wombats"** and got a letter from his boss that due to security reasons, from now on, the devs have to use the internal company registry for their packages. The internal registry can be accessed via username and password. The problem here arises that Jeff knows that in npm you can only set one main registry where to get the packages from.

Reviewer: `javscripty/typescript` is very wrong including a typo. Those names have to be camel case and are names: `JavaScript/TypeScript`. Those details matter as they tell how much you care for your words 🙏

```suggestion
Jeff Wombat Jr.[^1] is a Frontend developer at **"Working Wombats"** and got a letter from his boss that due to security reasons, from now on, the developers have to use the internal company registry for their npm packages. The internal registry can be accessed via username and password. The problem here arises that Jeff knows that in npm you can only set one main registry where to get the packages from.
```

[^1]: Imaginary person which does not reflect any persons in the real world.

## Scopes and scoped registries to the resuce

> All npm packages have a name. Some package names also have a scope. A scope follows the usual rules for package names (URL-safe characters, no leading dots or underscores). When used in package names, scopes are preceded by an `@` symbol and followed by a slash, e.g. `@somescope/somepackagename` Scopes are a way of grouping related packages together, and also affect a few things about the way npm treats the package.
> -- <cite>[npm](https://docs.npmjs.com/cli/v8/using-npm/scope)</cite>

> 

Reviewer: > The problem here arises that Jeff knows that in npm you can only set one main registry where to get the packages from.

Why is this a problem? You have something in front of your mind but you don't tell the reader.  I think you mean the problem with the local packages vs public packages. In addition you already suggest to know the core of the problem given by the boss directly in the next paragraph. That's like telling have of the story here. Why so fast?

Worse you then tell the solution in the headline:

> Scopes and scoped registries to the resuce

While I have expected a slow ramp up. A small explanation of scopes, maybe a link to the docs and then describe how that will help in the situation. Your story is upside down: "hey using scopes is the solution, here is why". That's like telling who is the murder in a book and then tell the story around it.

Author: I removed the immediate solution and put it more into the storyline what he does and then stumbling onto the scoped packages.

Each of these files is loaded, and config options are resolved in priority order. For example, a setting in the userconfig file would override the setting in the globalconfig file. So we could put the scoped registry into the userconfig file instead of the project config file.

## Summary

Packages in npm can be grouped together with scopes. Scopes are preceded by a `@` followed by the scope name and the package name separated by `/` e.g. `@scope/package`. To set a specific registry for a scope you can use scoped registries. To set a scoped registry open up the .npmrc file and enter the following:

```
@scope:registry=https://<your registry>
```

where `@scope` is your scope name and `<your registry>` is the registry where your scoped packages are hosted. 

If you need authentication you can add an additional line below the scoped registry with `_auth` and the basic authentication string for username and password (`username:password` ⇒ base64 encoded) or `_authToken` for an access token:

```
@scope:registry=https://<your registry>
//<your registry>:_auth=<base64-string>

> ```

Reviewer: Thanks for the summary. Is this meant as your conclusion too? I wrote exactly about that fact in my other review: The summary shouldn't be the conclusion. It still feels an abrupt ending.

Take the reader at the hand. Come back to `Jeff Wombat Jr.`. Tell that the boss is very happy with him or close the story with Jeff and say something general as the closing words like: npm scopes are crucial for enterprises. I hope I was able to give you a quick introduction to the topic. 

You can also include your personal motivation to write the blog post. I think the npm/scope story was new to you being mostly a backend developer. So why not telling that in the conclusion? There are many more options and all of them make the blog post more authentic and preventing an abrupt end.

Author: I updated the summary and put it into the story. Maybe the ending then is not so aprubpt. Also included something personal. Lets see if it reads better. :)

## Conclusion

Finally, Jeff could shut down his computer and leave work. Today he learned how to use scoped registries to access different registries for different scoped packages including how to use authentication. On his way home he quickly recapped what he has learned today:

Packages in npm can be grouped with scopes. Scopes are preceded by a `@` followed by the scope name and the package name separated by `/` e.g. `@scope/package`. To set a specific registry for a scope you can use scoped registries. To set a scoped registry open up the .npmrc file and enter the following:

```
@scope:registry=https://<your registry>
```

where `@scope` is the scope name and `<your registry>` is the registry where the scoped packages are hosted. 

If authentication is needed it is possible to add a line below the scoped registry with `_auth` and the basic authentication string for username and password (`username:password` ⇒ base64 encoded) or `_authToken` for an access token:

```
@scope:registry=https://<your registry>
//<your registry>:_auth=<base64-string>
```

> "Mhh", thought Jeff, in the end, this wasn't too hard. He put on his headphones and started hearing music and enjoyed the rest of the day.

Reviewer 3: Attention: Rant ahead.

> Finally Jeff could shutdown his computer and leave work. [...] On his way home he quickly recapped what he has learned today:
[...]
He put on his headphones and started hearing music and enjoyed the rest of the day.

This feels weird. Is he using the headphones with his phone after he left work? Also, while this storytelling approach to blog posts might be appealing to some, it's not to me. Just my opinion. It seems unprofessional to me to bloat a technical blog with a story that does not add value to the technical information.

Sidenote: I'm pretty sure "started listening to music" would be the proper translation here. "started hearing music" implies that all of a sudden there is music playing from somewhere unknown.

Reviewer: >  It seems unprofessional to me to bloat a technical blog with a story that does not add value to the technical information.

First of all: I think telling a story or not is a personal flavor of writing an article, it's not related to being professional or not. We are not talking about a user manual or technical documentation here. @Wompipomp is writing his first blog post so we should be open minded and not tear down the entire approach he chose 🙏

You could say the same not only about technology but also about anything that is transporting knowledge. I for myself like stories because it stretches the time you deal with a topic. This puts the topic in much deeper regions of my brain than just reading condensed facts. That's why I really don't like summary services like [blinkist](https://www.blinkist.com) for example.

Everyone has their own liking, I'm happy that you shared yours (and announced it as a rant to be prepared 😄). Mark could try out that writing style in his second blog post 🙌

---
path: "/blog/scoped-registry/"
date: "2021-12-30"

> title: "How to set up multiple npm registries with scoped registries"

Reviewer: This is my last review comment. I read everything and comment on a few things. Looking at the headline, I think this is not adequate. You are not exactly  telling a story about multiple npm registries.

The main problem is state in here:

>[Jeff] got a letter from his boss that due to security reasons, from now on, the developers have to use the internal company registry for their npm packages. The internal registry can be accessed via username and password. "This is an easy one," thinks Jeff and starts trying to add the new registry to the configuration. But soon he finds out, that he can only add one default registry. But how should he then get packages from the public npm registry and, on the other side, private packages from the newly setup internal registry at the same time? Time to google...

The underlying issue is about the common case that enterprises has an actual requirement for scoped registries:

```suggestion
title: "How enterprises benefit from scoped npm registries"
```

Author: Yes you are right that is way more catchy and describes more exactly what it is about :+1:

featuredImage: images/npm-scoped-registry.jpg
attribution:
    creator: Paul Esch-Laurent
    source: https://unsplash.com/photos/oZMUrWFHOB4

> seoMetaText: How to set up multiple npm registries with scoped packages and scoped registries even if they are protected

Reviewer: This need would need some more love depending on your decision to change the title or not.

Author: Updated the texts :+1: I think they are now way better.

teaserText: Learn with Jeff the Wombat, about npm scoped packages and scoped registries and how to configure access to multiple registries even if they are protected.
author: Mark Altmann
authorSummary: dev(ops)@satellytes, ❤️ for Wombats

---
## What is this all about?

> Nowadays almost every programming language comes with its package management solution, so developers can easily share packages with other developers around the world. For Python it is pip, for .NET it's NuGet. npm is the package manager solution for Javascript-based languages. It is used to create and use node packaged modules and is built into the Javascript platform [Node.js](http://www.nodejs.org/). The central component behind these package managers is a registry. A registry is a database of packages, each comprised of software and metadata. For example, the public registry for npm is [registry.npmjs.org](http://registry.npmjs.org).

Reviewer: It's JavaScript in camel case not Javascript☝️  I mentioned that before, maybe you can search and replace your post ?

Author: fixed :) Grammarly knows NuGet but not JavaScript :smile:  interesting

> Besides using public registries, companies can establish their private registries in their company network. [Nexus](https://www.sonatype.com/products/repository-pro), [Artifactory](https://jfrog.com/artifactory/) or [verdaccio](https://github.com/verdaccio/verdaccio) only to name some of the well-known. The advantage is that the published packages of that company never leave the company network. Another advantage is that you can set up authentication to additionally secure your internally published packages.

Reviewer: That reads great know and brings me much closer to the point of your blog post

Author: :raised_hands:

> Jeff the Wombat [^1] is a Frontend developer at **"Working Wombats"** and got a letter from his boss that due to security reasons, from now on, the developers have to use the internal company registry for their npm packages. The internal registry can be accessed via username and password. "This is an easy one," thinks Jeff and starts trying to add the new registry to the configuration. But soon he finds out, that he can only add one default registry. But how should he then get packages from the public npm registry and, on the other side, private packages from the newly setup internal registry at the same time? Time to google...

Reviewer: Nice that you try out the foot notes here. I never used them in markdown.

Author: I didn't even know if they are working but I was surprised that they look quite cool.

> Jeff the Wombat [^1] is a Frontend developer at **"Working Wombats"** and got a letter from his boss that due to security reasons, from now on, the developers have to use the internal company registry for their npm packages. The internal registry can be accessed via username and password. "This is an easy one," thinks Jeff and starts trying to add the new registry to the configuration. But soon he finds out, that he can only add one default registry. But how should he then get packages from the public npm registry and, on the other side, private packages from the newly setup internal registry at the same time? Time to google...

Reviewer: This feels like a hard cut here. You tell facts and introduce the topic and then mr. wombat comes around the corner with the problems. Maybe a headline "The challenges for enterprises" could help here? You can then merge the following section and remove head headline "Scopes and scoped registries". I have no other easy solution for that.

Author: Thats correct. I think I came up with a quite nice headline :raised_hands:

Reviewer: Yeah indeed! Thanks.

[^1]: Imaginary figure which does not reflect any certain person or animal in the real world.

## Scopes and scoped registries

While searching Jeff stumbles upon the following:

> All npm packages have a name. Some package names also have a scope. A scope follows the usual rules for package names (URL-safe characters, no leading dots or underscores). When used in package names, scopes are preceded by an `@` symbol and followed by a slash, e.g. `@somescope/somepackagename` Scopes are a way of grouping related packages together, and also affect a few things about the way npm treats the package.
> -- <cite>[npm](https://docs.npmjs.com/cli/v8/using-npm/scope)</cite>

Packages can be grouped with scopes and for scopes, we can set alternative registries named `scoped registries` where npm will lookup the packages instead of the main registry.

"Well, this sounds promising!" Jeff thinks.

> 💡 The usage of scoped packages and scoped registries can also greatly diminish your attack surface for the dependency confusion attack [dependency confusion attack](https://snyk.io/blog/detect-prevent-dependency-confusion-attacks-npm-supply-chain-security/)

> ## Where to configure scoped registries

Reviewer: ```suggestion
## How to configure scoped registries
```

> .npmrc is a configuration file that configures how npm works in your environment. It can be used globally or on a user or project level. [Here](https://docs.npmjs.com/cli/v8/configuring-npm/npmrc) are the docs for .npmrc.

Reviewer: I appreciate the link to the official docs 👍

> **Working wombats** already set up the new registry and moved all packages from the public registry into the private registry. Fortunately for the internal packages, they already used scopes (`@workingwombats`). The registry can be found internally via the URL `registry.working-wombats.rocks`. So Jeff could add a scoped registry with the following line to .npmrc:

Reviewer: You refer to the company here which is not clear as you chose to write it different than before "Working wombats" vs "Working Wombats". I would also add [ACME](https://en.wikipedia.org/wiki/Acme_Corporation) to the company name as it makes it super easy to understand that it's a fake company (see the linked article).  So replace every name of the company with "Working Wombats ACME"

Author: :+1: ah that's really nice. So we don't have problems with real company names :rabbit: :raised_hands:

`sirWombat:theCudd1er!` ⇒ base64 encoded ⇒ `c2lyV29tYmF0OnRoZUN1ZGQxZXIh`

So to add basic authentication to the registry, Jeff encodes the username and password with base64 and goes back to the .npmrc file, and adds the following line below the corresponding scoped registry:

```
@workingwombats:registry=https://registry.working-wombats.rocks/
//registry.working-wombats.org/:_auth=c2lyV29tYmF0OnRoZUN1ZGQxZXIh
```

>💡 Note the // without the HTTPS: at the beginning

Alternatively, if you have an access token you can put the access token after `_authToken`:

```
@workingwombats:registry=https://registry.working-wombats.rocks/
//registry.working-wombats.org/:_authToken=c2lyd29tYmF0OnRoZWN1ZGQxZXIh
```

Jeff tried it again et voilà it worked! He almost committed the new additions of the project-level .npmrc to git but then his old mentor "Sir Wombat" came into his mind:

>💡 Remember that it is not a good practice to add credentials to version control e.g. if you have versioned your .npmrc file.

and Jeff was asking himself...

> ## But where should I put the credentials then? 🤷‍♂️

Reviewer: Relating to headlines from a paragraph is for myself very confusing. I also don't see that very often so I think it's probably not ideal. Let's keep it for this blog post👍

If you have a project-level .npmrc file and have added it in your version control then you could do the following:

**Use environment variables**

> we could export the authentication string as an environment variable:

Reviewer: Looks awkward I would just write a normal sentence and emphasize that part in there:

```suggestion
Jeff doesn't want to commit the credentials so he decides to put the credentials in an **environment variable** and refer to them in the configuration file. 

Start by defining the variable:
```

Author: ah ok. yes, I thought that I bring up at least 2 possibilities how to not store the credentials in git so I did this as kind of list. But with a normal list I had problems with code blocks etc. :smile:

e.g. for linux `$ export NPM_TOKEN=<base64-string>`

> We can then add the env variable in our .npmrc file:

Reviewer: ```suggestion
We can then add the environment variable in our .npmrc file:
```

People will understand it but introducing abbreviations just to save writing time is not good. I would always aim to write the whole word.

Author: Yeah I missed that one. Thats a good point

```
//registry.wombatcorp.org/:_auth=${NPM_TOKEN}
```

> The npm CLI will replace this value with the contents of the NPM_TOKEN environment variable

Reviewer: You are missing a period.
```suggestion
The npm CLI will replace this value with the contents of the NPM_TOKEN environment variable.
```

Author: :raised_hands:

> **Add scoped registry to user-level instead of project-level**

Reviewer: I think you can remove the headline (which is not a real headline anyway). The following text makes sense without it.

Author: with the change above then definitely :grin:

Author: I removed the whole paragraph about different npmrc files. Like I said above I wanted to emphasize that there are more possibilities how to not save the credentials in git and, to not leave the reader alone searching for solutions, at least, put 2 possibilities there. But the second block sounds a bit out of context if we remove the first headline. And if you didn't understand the listing then the usual reader won't understand either :smile: so let's go with one possibility what jeff does because this is also valuable information that npm replaces environment variables.

There are different .npmrc files as explained here ([npmrc](https://docs.npmjs.com/cli/v8/configuring-npm/npmrc)). 
Each of these files is loaded, and config options are resolved in priority order. For example, a setting in the user config file would override the setting in the global config file. So we could put the scoped registry into the user config file instead of the project config file.

> ## Conclusion

Reviewer: Thank you for the conclusion. Rounds it up much better than a summary which you merged buy including a recap. I for myself would remove the technical recap/summary because your blog post is very short so people don't need to read again about that here.

I would like to hear about Jeff, that he learned a few things about scopes, tokens and npmrc files and that it wasn't super hard. Done.

Author: :+1:

> Finally, Jeff could shut down his computer and leave work. Today he learned how to use scoped registries to access different scoped packages including how to use authentication. On his way home he quickly recapped what he has learned today:

Reviewer: I for myself struggle to maintain a correct tense when writing English. This feels wrong:

> Finally, Jeff could shut down his computer and leave work. 

vs (as the story is currently telling something about Jeff, so it's present tense))

> Finally, Jeff can shut down his computer and leave work. 

but if this makes sense then you would have to check the tenses through the whole article again.
